geopy/geocoders/geoclient.py

```diff
@@ -0,0 +1,156 @@
+                )
+        
+        if not app_key:
+            raise ConfigurationError(
+                'No app_key given, required for api access.  If you do not '
+                'have a Geoclient app_key, sign up here: '
+                'https://developer.cityofnewyork.us/api/geoclient-api'
+                )
+        
+        self.domain = domain.strip('/')
+        
+        '''use SFS (Single-field search) endpoint'''
+        self.geocode_api = 'https://%s/search.json' % (self.domain)
+
+    def geocode(
+            self,
+            query,
+            exactly_one=True,
+            timeout=None,
+            addressdetails=False,
+        ):
+        """
+        Geocode a location query.
+
+        :param string query: The query string to be geocoded using SFS syntax;
+            this must be URL encoded. Documentation at:
+            https://api.cityofnewyork.us/geoclient/v1/doc#section-1.3
+
+        :param bool exactly_one: Return one result or a list of results, if
+            available.
+
+        :param int timeout: Time, in seconds, to wait for the geocoding service
+            to respond before raising a :class:`geopy.exc.GeocoderTimedOut`
+            exception. Set this only if you wish to override, on this call
+            only, the value set during the geocoder's initialization.
+
+        :param bool addressdetails: If you want in *Location.raw* to include
+            address details such as BIN#, community district, etc set it to True.
+            Data Dictionary documenting all available fields at:
+            https://api.cityofnewyork.us/geoclient/v1/doc#section-3.0
+
+        """
+        params = {
+            'input': self.format_string % query,
+            'app_key': self.app_key,
+            'app_id': self.app_id,
+        }
+
+        url = "?".join((self.geocode_api, urlencode(params)))
+
+        logger.debug("%s.geocode: %s", self.__class__.__name__, url)
+        return self._parse_json_geocode(
+            self._call_geocoder(url, timeout=timeout), exactly_one, addressdetails
+        )
+
+
+    @staticmethod
+    def _parse_json_geocode(page, exactly_one=True, addressdetails=False):
+        '''Returns location from Geoclient response.'''
+
+        places = page['results']
+
+        if not len(places):
+            return None
+
+        def parse_place(place):
+            '''Get the location, lat, lon from a single json result.'''
+            '''Return normalized street address in location, and the entire Geoclient response is also returned'''
+            '''if addressdetails is set to true should the user want to parse extended location information'''
+            
+            request_type = place.get('request').split(' ', 1)[0]
+            
+            response_info = place['response']
+            
+            if request_type == 'address':
+                location = response_info.get('houseNumber') + ' ' + \
```

Comment:
Are these truly optional? If so we want `response_info.get('houseNumber', '')` or another default value, I'd think, to avoid serializing `None`? Also with this many arguments `', '.join((...))` would be good for performance.